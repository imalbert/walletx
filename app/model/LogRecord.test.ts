import {
  getSnapshot, IJsonPatch, onPatch,
} from 'mobx-state-tree'

import {
  Log,
  Record,
  LogModelType,
  LOG_TYPES,
  LOG_CATEGORY,
  RECORD_ACCOUNT,
  RecordModelType,
} from './LogRecord'
import { autorun, reaction } from 'mobx'

const DEFAULT_DATE = new Date(1581819204151)
const DEFAULT_BALANCE = 1000
const DEFAULT_ACCOUNT = 'CASH_IN_HAND'
const DEFAULT_LOG_AMOUNT = 500
const DEFAULT_LOG_CATEGORY = LOG_CATEGORY.FOOD
const DEFAULT_LOG_TYPE = LOG_TYPES.EXPENSE
const buildLogObj = (overrides?): LogModelType => ({
  amount: DEFAULT_LOG_AMOUNT,
  category: DEFAULT_LOG_CATEGORY,
  date: DEFAULT_DATE,
  type: DEFAULT_LOG_TYPE,
  ...overrides,
})

describe('Log Model', () => {

  it('can create an instance', () => {
    const log = Log.create(buildLogObj())

    expect(getSnapshot(log)).toMatchSnapshot()
  })

  describe('actions', () => {
    let log: LogModelType
    let patches: Array<IJsonPatch>

    beforeEach(() => {
      patches = []
      log = Log.create(buildLogObj())

      onPatch(log, snapshot => patches.push(snapshot))
    })
    afterEach(() => { expect(patches).toMatchSnapshot() })

    it('can update a log', () => {
      log.changeAmount(999)
      log.changeCategory(LOG_CATEGORY.HOUSEHOLD)
      log.changeDate(new Date(1581819204152))
      log.changeType(LOG_TYPES.INCOME)

      expect(getSnapshot(log)).toMatchSnapshot()
    })
  })
})

describe('Records Model', () => {
  it('can create a log record with defaults', () => {
    const record_with_defaults = Record.create()

    expect(getSnapshot(record_with_defaults)).toMatchSnapshot()
  })

  it('can create a log record with defaults', () => {
    const record_with_props = Record.create({
      balance: DEFAULT_BALANCE,
      logs: [Log.create(buildLogObj())],
    })

    expect(getSnapshot(record_with_props)).toMatchSnapshot()
  })

  describe('views', () => {
    test('getLogsByMonth to return array of logs by month & year', () => {
      let logsByMonth = []
      const record = Record.create({})
      autorun(() => {
        logsByMonth = record.getLogsByMonth(new Date(2020, 1, 20)) // February 2020
        console.log(logsByMonth)
      })

      record.add(Log.create({ amount: 1, date: new Date('February 1, 2020 03:24:00') }))
      record.add(Log.create({ amount: 2, date: new Date('January 1, 2020 03:24:00') }))
      record.add(Log.create({ amount: 3, date: new Date(2020, 1, 1) }))
      record.add(Log.create({ amount: 4, date: new Date(2020, 2, 1) }))
      record.add(Log.create({ amount: 5, date: new Date(2021, 8, 10) }))
      record.add(Log.create({ amount: 6, date: new Date(2020, 1, 14) }))
      // expect 1, 3, 6 (February and 2020)

      expect(getSnapshot(record)).toMatchSnapshot()
      expect(logsByMonth).toMatchSnapshot()
    })

    test('getLogsByDayOfMonth to return map of logs with (month + day) as property', () => {
      const record = Record.create({})

      record.add(Log.create({ amount: 1, date: new Date('February 1, 2020 03:24:00') }))
      record.add(Log.create({ amount: 2, date: new Date('January 1, 2020 03:24:00') }))
      record.add(Log.create({ amount: 3, date: new Date(2020, 1, 1) }))
      record.add(Log.create({ amount: 4, date: new Date(2020, 2, 1) }))
      record.add(Log.create({ amount: 5, date: new Date(2021, 8, 10) }))
      record.add(Log.create({ amount: 6, date: new Date(2020, 1, 14) }))

      expect(record.getLogsByDayOfMonth(new Date(2020, 1, 1))).toMatchSnapshot()
      expect(record.getLogsByDayOfMonth(new Date(2021, 8, 1))).toMatchSnapshot()
      expect(record.getLogsByDayOfMonth(new Date(2020, 2, 1))).toMatchSnapshot()
    })
  })

  describe('actions', () => {
    let record: RecordModelType
    let patches: Array<IJsonPatch>

    beforeEach(() => {
      patches = []
      record = Record.create({ balance: DEFAULT_BALANCE })

      onPatch(record, snapshot => patches.push(snapshot))
    })
    afterEach(() => { expect(patches).toMatchSnapshot() })

    it('can record a log', () => {
      record.add(Log.create(buildLogObj()))
      record.add(Log.create(buildLogObj({ type: LOG_TYPES.INCOME })))
    })

    it('can remove all logs', () => {
      record.add(Log.create(buildLogObj()))
      record.add(Log.create(buildLogObj()))
      const totalExpenses = DEFAULT_LOG_AMOUNT * 2

      expect(record.logs.length).toBe(2)
      expect(record.balance).toBe(DEFAULT_BALANCE - totalExpenses)

      record.clear()
      expect(record.logs.length).toBe(0)
      expect(record.balance).toBe(DEFAULT_BALANCE)
    })
  })
})
