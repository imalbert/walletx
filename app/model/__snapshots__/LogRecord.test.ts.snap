// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log Model actions can update a log 1`] = `
Object {
  "amount": 999,
  "category": "INCOME",
  "date": 1581819204152,
  "type": "INCOME",
}
`;

exports[`Log Model actions can update a log 2`] = `
Array [
  Object {
    "op": "replace",
    "path": "/amount",
    "value": 999,
  },
  Object {
    "op": "replace",
    "path": "/category",
    "value": "HOUSEHOLD",
  },
  Object {
    "op": "replace",
    "path": "/date",
    "value": 1581819204152,
  },
  Object {
    "op": "replace",
    "path": "/type",
    "value": "INCOME",
  },
  Object {
    "op": "replace",
    "path": "/category",
    "value": "INCOME",
  },
]
`;

exports[`Log Model can create an instance 1`] = `
Object {
  "amount": 500,
  "category": "FOOD",
  "date": 1581819204151,
  "type": "EXPENSE",
}
`;

exports[`Records Model actions can record a log 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/logs/0",
    "value": Object {
      "amount": 500,
      "category": "FOOD",
      "date": 1581819204151,
      "type": "EXPENSE",
    },
  },
  Object {
    "op": "add",
    "path": "/logs/1",
    "value": Object {
      "amount": 500,
      "category": "FOOD",
      "date": 1581819204151,
      "type": "INCOME",
    },
  },
]
`;

exports[`Records Model actions can remove all logs 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/logs/0",
    "value": Object {
      "amount": 1000,
      "category": "FOOD",
      "date": 1581819204151,
      "type": "INCOME",
    },
  },
  Object {
    "op": "add",
    "path": "/logs/1",
    "value": Object {
      "amount": 500,
      "category": "FOOD",
      "date": 1581819204151,
      "type": "EXPENSE",
    },
  },
  Object {
    "op": "add",
    "path": "/logs/2",
    "value": Object {
      "amount": 500,
      "category": "FOOD",
      "date": 1581819204151,
      "type": "EXPENSE",
    },
  },
  Object {
    "op": "remove",
    "path": "/logs/0",
  },
  Object {
    "op": "remove",
    "path": "/logs/0",
  },
  Object {
    "op": "remove",
    "path": "/logs/0",
  },
]
`;

exports[`Records Model can create a log record with defaults 1`] = `
Object {
  "account": "CASH_IN_HAND",
  "logs": Array [],
}
`;

exports[`Records Model can create a log record with defaults 2`] = `
Object {
  "account": "CASH_IN_HAND",
  "logs": Array [
    Object {
      "amount": 500,
      "category": "FOOD",
      "date": 1581819204151,
      "type": "EXPENSE",
    },
  ],
}
`;

exports[`Records Model views getLogsByDayOfMonth to return map of logs with (month + day) as property 1`] = `
Object {
  "Feb 01": Array [
    Object {
      "amount": 1,
      "category": "FOOD",
      "date": 1580498640000,
      "type": "EXPENSE",
    },
    Object {
      "amount": 3,
      "category": "FOOD",
      "date": 1580486400000,
      "type": "EXPENSE",
    },
  ],
  "Feb 14": Array [
    Object {
      "amount": 6,
      "category": "FOOD",
      "date": 1581609600000,
      "type": "EXPENSE",
    },
  ],
}
`;

exports[`Records Model views getLogsByDayOfMonth to return map of logs with (month + day) as property 2`] = `
Object {
  "Sep 10": Array [
    Object {
      "amount": 5,
      "category": "FOOD",
      "date": 1631203200000,
      "type": "EXPENSE",
    },
  ],
}
`;

exports[`Records Model views getLogsByDayOfMonth to return map of logs with (month + day) as property 3`] = `
Object {
  "Mar 01": Array [
    Object {
      "amount": 4,
      "category": "FOOD",
      "date": 1582992000000,
      "type": "EXPENSE",
    },
  ],
}
`;

exports[`Records Model views getLogsByMonth to return array of logs by month & year 1`] = `
Object {
  "account": "CASH_IN_HAND",
  "logs": Array [
    Object {
      "amount": 1,
      "category": "FOOD",
      "date": 1580498640000,
      "type": "EXPENSE",
    },
    Object {
      "amount": 2,
      "category": "FOOD",
      "date": 1577820240000,
      "type": "EXPENSE",
    },
    Object {
      "amount": 3,
      "category": "FOOD",
      "date": 1580486400000,
      "type": "EXPENSE",
    },
    Object {
      "amount": 4,
      "category": "FOOD",
      "date": 1582992000000,
      "type": "EXPENSE",
    },
    Object {
      "amount": 5,
      "category": "FOOD",
      "date": 1631203200000,
      "type": "EXPENSE",
    },
    Object {
      "amount": 6,
      "category": "FOOD",
      "date": 1581609600000,
      "type": "EXPENSE",
    },
  ],
}
`;

exports[`Records Model views getLogsByMonth to return array of logs by month & year 2`] = `
Array [
  Object {
    "amount": 1,
    "category": "FOOD",
    "date": 1580498640000,
    "type": "EXPENSE",
  },
  Object {
    "amount": 3,
    "category": "FOOD",
    "date": 1580486400000,
    "type": "EXPENSE",
  },
  Object {
    "amount": 6,
    "category": "FOOD",
    "date": 1581609600000,
    "type": "EXPENSE",
  },
]
`;
